optimum/configuration_utils.py

```diff
@@ -151,114 +187,153 @@ def get_config_dict(
-        is_local = os.path.isdir(pretrained_model_name_or_path)
 
-        if os.path.isfile(pretrained_model_name_or_path):
+        is_local = os.path.isdir(pretrained_model_name_or_path)
+        if os.path.isfile(os.path.join(subfolder, pretrained_model_name_or_path)):
+            # Special case when pretrained_model_name_or_path is a local file
             resolved_config_file = pretrained_model_name_or_path
             is_local = True
-        elif os.path.isdir(pretrained_model_name_or_path):
+        # TODO: remove once transformers release version is way above 4.22.
+        elif _transformers_version_is_below_threshold and os.path.isdir(pretrained_model_name_or_path):
+            configuration_file = kwargs.pop("_configuration_file", cls.CONFIG_NAME)
             resolved_config_file = os.path.join(pretrained_model_name_or_path, configuration_file)
             if not os.path.isfile(resolved_config_file):
                 raise EnvironmentError(
                     f"Could not locate {configuration_file} inside {pretrained_model_name_or_path}."
                 )
+        # TODO: remove condition once transformers release version is way above 4.22.
+        elif not _transformers_version_is_below_threshold and is_remote_url(pretrained_model_name_or_path):
+            configuration_file = pretrained_model_name_or_path
+            resolved_config_file = download_url(pretrained_model_name_or_path)
         else:
+            configuration_file = kwargs.pop("_configuration_file", cls.CONFIG_NAME)
+
             try:
-                resolved_config_file = hf_hub_download(
-                    repo_id=pretrained_model_name_or_path,
-                    filename=configuration_file,
-                    revision=revision,
-                    cache_dir=cache_dir,
-                    force_download=force_download,
-                    resume_download=resume_download,
-                    proxies=proxies,
-                    use_auth_token=use_auth_token,
-                    local_files_only=local_files_only,
-                )
+                # TODO: remove once transformers release version is way above 4.22.
+                if _transformers_version_is_below_threshold:
+                    resolved_config_file = hf_hub_download(
+                        repo_id=pretrained_model_name_or_path,
+                        filename=configuration_file,
+                        revision=revision,
+                        cache_dir=cache_dir,
+                        force_download=force_download,
+                        resume_download=resume_download,
+                        proxies=proxies,
+                        use_auth_token=use_auth_token,
+                        local_files_only=local_files_only,
+                    )
+                else:
+                    # Load from local folder or from cache or download from model Hub and cache
+                    resolved_config_file = cached_file(
```

Comment:
We should either use `cached_path` and `cached_file` (depending on which transformers version) or use `hf_hub_download` , I prefer the second option personnaly

Reply:
I changed, and use `cached_file` to match what was done for transformers <= 4.21.0
For the rest, I think it is better to match the `transformers` implementation since `BaseConfig` is just a customizable `transformers.PretrainedConfig`. We actually want to copy anything related to pushing to hub from this class, so we should use their implementation when it is possible, and here it is.

Reply:
I don't see the gain in using `cached_path` and `cached_file` : our current implementation  with `hf_hub_download` is equivalent so I would be in favor of the option where we don't have additional statements to check which transformers version is used (option which will be more difficult to maintain)

Reply:
- `cached_path` is from the implementation from `transformers <= 4.21.0` and we are just matching the base class we are extending here
- `cached_file` uses `hf_hub_download` underneath, and does other stuff related to the offline mode and loading from the cache with a commit hash (which seems to be new). Once again, `BaseConfig` inherints from`transformers.PretrainedConfig`, so matching its behaviour will prevent issues: what if someday `cached_file` is further updated? Here we don't have a constraint on what to use, so we should use what's identical from what we are basing our code.

About the complexity, we will remove those if / else statements in a few release, once transformers 4.22 has been well adopted.

Reply:
I would be in favor of removing the `PretrainedConfig` inheritence as some modifications in `PretrainedConfig` could potentially lead to some unexpected behavior and tracking every changes would be time consuming. I am ok with merging https://github.com/huggingface/optimum/pull/386 in the meantime